pkg/controllers/applicationset_controller.go

```diff
@@ -100,6 +120,68 @@ func (r *ApplicationSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
 	}, nil
 }
 
+// Return true if there are unknown generators specified in the application set.  If we can discover the names
+// of these generators, return the names as the keys in a map
+func invalidateGenerators(applicationSetInfo *argoprojiov1alpha1.ApplicationSet) (bool, map[string]bool) {
+	names := make(map[string]bool)
+	hasInvalidGenerators := false
+	for index, generator := range applicationSetInfo.Spec.Generators {
+		v := reflect.Indirect(reflect.ValueOf(generator))
+		found := false
+		for i := 0; i < v.NumField(); i++ {
+			field := v.Field(i)
+			if !field.CanInterface() {
+				continue
+			}
+			if !reflect.ValueOf(field.Interface()).IsNil() {
+				found = true
+				break
+			}
+		}
+		if !found {
+			hasInvalidGenerators = true
+			addInvalidGeneratorNames(names, applicationSetInfo, index)
+		}
+	}
+	return hasInvalidGenerators, names
+}
+
+func addInvalidGeneratorNames(names map[string]bool, applicationSetInfo *argoprojiov1alpha1.ApplicationSet, index int) {
+	// The generator names are stored in the "kubectl.kubernetes.io/last-applied-configuration" annotation
+	config := applicationSetInfo.ObjectMeta.Annotations["kubectl.kubernetes.io/last-applied-configuration"]
+	var values map[string]interface{}
+	err := json.Unmarshal([]byte(config), &values)
+	if err != nil {
+		log.Warnf("couldn't unmarshal kubectl.kubernetes.io/last-applied-configuration: %+v", config)
+		return
+	}
+
+	spec, ok := values["spec"].(map[string]interface{})
+	if !ok {
+		log.Warnf("coundn't get spec from kubectl.kubernetes.io/last-applied-configuration annotation")
+		return
+	}
+
+	generators, ok := spec["generators"].([]interface{})
+	if !ok {
+		log.Warnf("coundn't get generators from kubectl.kubernetes.io/last-applied-configuration annotation")
+		return
+	}
+
+	generator, ok := generators[index].(map[string]interface{})
+	if !ok {
+		log.Warnf("coundn't get generator from kubectl.kubernetes.io/last-applied-configuration annotation")
+		return
+	}
+
+	var name string
+	for key := range generator {
+		name = key
+		names[name] = true
```

Comment:
Nitpick: think this can just be `names[key] = true`

Reply:
Done